pkg/trace/writer/newwriter/sender_test.go

```diff
@@ -0,0 +1,238 @@
+package writer
+
+import (
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"os"
+	"strconv"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/cihub/seelog"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestMain(m *testing.M) {
+	log.SetupDatadogLogger(seelog.Disabled, "error")
+	os.Exit(m.Run())
+}
+
+func TestSender(t *testing.T) {
+	const climit = 100
+	testSenderConfig := func(serverURL string) *senderConfig {
+		url, err := url.Parse(serverURL + "/")
+		if err != nil {
+			t.Fatal(err)
+		}
+		return &senderConfig{
+			client:   &http.Client{},
+			url:      url,
+			maxConns: climit,
+			apiKey:   "123",
+		}
+	}
+
+	t.Run("accept", func(t *testing.T) {
+		assert := assert.New(t)
+		server := newTestServer()
+		defer server.Close()
+
+		s := newSender(testSenderConfig(server.URL))
+		for i := 0; i < 20; i++ {
+			s.Push(expectResponses(200))
+		}
+		s.Flush()
+
+		assert.Equal(20, server.Total(), "total")
+		assert.Equal(20, server.Accepted(), "accepted")
+		assert.Equal(0, server.Retried(), "retry")
+		assert.Equal(0, server.Failed(), "failed")
+	})
+
+	t.Run("accept-many", func(t *testing.T) {
+		assert := assert.New(t)
+		server := newTestServer()
+		defer server.Close()
+
+		s := newSender(testSenderConfig(server.URL))
+		for i := 0; i < climit*4; i++ {
+			s.Push(expectResponses(200))
```

Comment:
I feel like we could have another test ensuring payloads are processed concurrently according to climit. I don't think any of the existing ones does this.

Reply:
Yes, it is definitely on my list to add more tests. I just stopped for a bit to get some initial reviews, to make sure I haven't omitted anything in the overall logic, in case someone reveals some drastic changes that will be needed which will in turn require changing more tests.

Reply:
I added another commit with a test for this, but I'm not sure it's 100% correct.